### ISSUE TITLE
Heatmap tensor visualization showing no information

### ISSUE THREAD

**Issue by shivahanifi** [shivahanifi/GazeMDETR#5]
To see the difference between the resized image form of the grayscale norm_map and downsampled norm_map tensor I plotted both using the following code:

## Roadmap

![heatmap_evolution](https://github.com/shivahanifi/GazeMDETR/assets/78621902/6b9e9303-b942-4823-b330-fe69af61a9c5)

## Code
```
# norm_map normalize and resize as image
normalized_norm_map = norm_map_gray.point(lambda x: (x - 127.5) / 127.5)
transform_normMap_noTensor = T.transforms.Compose([
    T.Resize(25)
])
reszied_normmap_image = transform_normMap_noTensor(normalized_norm_map)
print("reszied_normmap_image max: ", max(reszied_normmap_image.getdata()), "\n reszied_normmap_image shape: ", reszied_normmap_image.size)

# norm_map normalize and reszie as tensor
transform_normMap = T.transforms.Compose([
    T.Resize(800),
    T.transforms.ToTensor()
])
normalized_norm_map_tensor = transform_normMap(norm_map_gray)

# Visualize downsampled norm map
downsampled_norm_map = torch.nn.functional.interpolate(normalized_norm_map_tensor.unsqueeze(0),size=(25,34), mode='bilinear', align_corners=False).squeeze(0)
print("downsampled_norm_map max: ", torch.max(downsampled_norm_map), "\n downsampled_norm_map shape: ", downsampled_norm_map.shape)
downsampled_norm_map_array = np.squeeze(downsampled_norm_map.cpu().numpy().astype(np.uint8))
downsampled_norm_map_image = Image.fromarray(downsampled_norm_map_array)

fig, axs = plt.subplots(1, 2, figsize=(10, 5))
axs[0].imshow(reszied_normmap_image)
axs[0].set_title("downsampled heatmap image")

axs[1].imshow(downsampled_norm_map_image)
axs[1].set_title("downsampled heatmap tensor")
plt.show()
```

## Output

|variable name|size/shape|notes|
|---|---|---|
|reszied_normmap_image|(33, 25)|max value 1|
|normalized_norm_map_tensor|torch.Size([1, 800, 1066])|max value 1#|
|normalized_norm_map_tensor_array|(800, 1066)|![normalized_norm_map](https://github.com/shivahanifi/GazeMDETR/assets/78621902/a9117e02-04b6-41a2-b6f1-9a8dc339f769)|
| downsampled_norm_map|torch.Size([1, 25, 34])|max value 0.9625/ bilinear interpolation|
|downsampled_norm_map_array |(25, 34)||

![resized_vs_downsampled_heatmap](https://github.com/shivahanifi/GazeMDETR/assets/78621902/e8910982-71c2-49d0-9cc1-036a886e943e)

https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#interpolate

**Comment by shivahanifi** [comment]
## Different interpolation modes for downsampling the tensor
|bilinear|area|nearest|nearest-exact|
|---|---|---|---|
|![interpolation_bilinear](https://github.com/shivahanifi/GazeMDETR/assets/78621902/43ea2e1b-31e7-400b-ada0-b7f2b60266dd)|![interpolation_area](https://github.com/shivahanifi/GazeMDETR/assets/78621902/528e9ea0-d03e-4946-94e6-bbb1093ba604)|![interpolation_nearest](https://github.com/shivahanifi/GazeMDETR/assets/78621902/c39a8d20-28eb-4812-afcc-9e633a8323a7)|![interpolation_nearest_exact](https://github.com/shivahanifi/GazeMDETR/assets/78621902/33b3fb6e-8f28-42b1-a3d1-5c7bdc202577)|

**Comment by shivahanifi** [comment]
## Problem solved
`.astype(np.uint8)` was the problem behind having tensor visualizations in all black. Previously I was converting tensor data to int before having them multiplied with 255. Therefore, all the values were converted to 0 since it was the closest integer. The following code fixed the problem.

### Code
```
# norm_map normalize and resize
normalized_norm_map = norm_map_gray.point(lambda x: (x - 127.5) / 127.5)
transform_normMap_noTensor = T.transforms.Compose([
    T.Resize(size=(25,34))
])
reszied_normmap_image = transform_normMap_noTensor(normalized_norm_map)
print("reszied_normmap_image max: ", max(reszied_normmap_image.getdata()), "\n reszied_normmap_image shape: ", reszied_normmap_image.size)


# resized image to tensor
transform_normMap_image_Tensor = T.transforms.Compose([
    T.transforms.ToTensor()
])
reszied_normmap_image_tensor = transform_normMap_image_Tensor(reszied_normmap_image)
reszied_normmap_image_tensor_array = np.squeeze((reszied_normmap_image_tensor.cpu().numpy())*255)
print("normalized_norm_map_tensor_array shape", reszied_normmap_image_tensor.shape)
reszied_normmap_image_tensor_image = Image.fromarray(reszied_normmap_image_tensor_array.astype(np.uint8))
reszied_normmap_image_tensor_image.show()

# Normalize and reszie as tensor
transform_normMap = T.transforms.Compose([
    T.Resize(800),
    T.transforms.ToTensor()
])
normalized_norm_map_tensor = transform_normMap(norm_map_gray)
print("normalized_norm_map_tensor max: ", torch.max(normalized_norm_map_tensor), "\n normalized_norm_map_tensor shape: ", normalized_norm_map_tensor.shape, "\n",normalized_norm_map_tensor)

# Visualize norm_map tensor before downsampling
normalized_norm_map_tensor_array = np.squeeze(normalized_norm_map_tensor.cpu().numpy())*255
#print("normalized_norm_map_tensor_array shape", normalized_norm_map_tensor_array.shape)
normalized_norm_map_tensor_image = Image.fromarray(normalized_norm_map_tensor_array.astype(np.uint8))
normalized_norm_map_tensor_image.show()

# Visualize downsampled norm map
downsampled_norm_map = torch.nn.functional.interpolate(normalized_norm_map_tensor.unsqueeze(0),size=(25,34), mode='bilinear', align_corners=False).squeeze(0)
print("downsampled_norm_map max: ", torch.max(downsampled_norm_map), "\n downsampled_norm_map shape: ", downsampled_norm_map.shape)
downsampled_norm_map_array = np.squeeze(downsampled_norm_map.cpu().numpy())*255
downsampled_norm_map_image = Image.fromarray(downsampled_norm_map_array.astype(np.uint8))

fig, axs = plt.subplots(1, 2, figsize=(10, 5))
axs[0].imshow(reszied_normmap_image, cmap='gray')
axs[0].set_title("resized heatmap image")

axs[1].imshow(downsampled_norm_map_image, cmap='gray')
axs[1].set_title("downsampled heatmap tensor")
plt.show()
```

### Output

- normalized_norm_map_tensor

![normalized_norm_map_tensor](https://github.com/shivahanifi/GazeMDETR/assets/78621902/a5406634-6f4a-4764-afba-37d5ad3f95fe)

![downsampledHeatmapTensor_visualization](https://github.com/shivahanifi/GazeMDETR/assets/78621902/aecd5502-0501-4a20-b0b8-80d4b80f56f8)
